File: app/src/main/java/org/bspb/smartbirds/pro/ui/views/SingleChoiceRadioFormInput.java
package org.bspb.smartbirds.pro.ui.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EViewGroup;
import org.androidannotations.annotations.ViewById;
import org.bspb.smartbirds.pro.R;
import org.bspb.smartbirds.pro.ui.exception.ViewValidationException;
import org.bspb.smartbirds.pro.ui.utils.FormsConfig;

import java.util.Map;

import retrofit2.http.POST;

@EViewGroup(R.layout.single_choice_radio)
public class SingleChoiceRadioFormInput extends FrameLayout implements SupportRequiredView, SupportStorage {

    private boolean mRequired;
    private CharSequence mEntriesKey;
    private CharSequence mHint;

    @ViewById(R.id.single_choice_radio_hint)
    protected TextView hintTextView;
    @ViewById(R.id.single_choice_radio_group)
    protected RadioGroup radioGroup;
    private FormsConfig mConfig;


    public SingleChoiceRadioFormInput(Context context) {
        this(context, null);
    }

    public SingleChoiceRadioFormInput(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.singleChoiceRadioFormInputStyle);
    }

    public SingleChoiceRadioFormInput(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.SingleChoiceRadioFormInput, defStyleAttr, 0);
        try {
            mHint = a.getText(R.styleable.SingleChoiceRadioFormInput_hint);
            mEntriesKey = a.getText(R.styleable.SingleChoiceRadioFormInput_config_entries);
            mRequired = a.getBoolean(R.styleable.SingleChoiceRadioFormInput_required, false);
        } finally {
            a.recycle();
        }
    }

    @AfterViews
    protected void init() {
        radioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                if (checkedId != -1) {
                    ((RadioButton) group.getChildAt(group.getChildCount() - 1)).setError(null);
                }
            }
        });

        if (mHint != null && mHint.length() > 0) {
            hintTextView.setText(mHint + ":");
        }

        if (mEntriesKey == null) {
            throw new IllegalArgumentException("Entries are required for this form input.");
        }

        mConfig = FormsConfig.valueOf(mEntriesKey.toString());

        for (int i = 0; i < mConfig.getValues().length; i++) {
            RadioButton radioButton = new RadioButton(getContext());
            radioButton.setText(mConfig.getLabels()[i]);
            radioGroup.addView(radioButton, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        }
    }

    @Override
    public void checkRequired() throws ViewValidationException {
        if (mRequired && isEnabled()) {
            if (radioGroup.getCheckedRadioButtonId() == -1) {
                ((RadioButton) radioGroup.getChildAt(radioGroup.getChildCount() - 1)).setError(getContext().getString(R.string.required_field));
                throw new ViewValidationException();
            } else {
                ((RadioButton) radioGroup.getChildAt(radioGroup.getChildCount() - 1)).setError(null);
            }
        } else {
            ((RadioButton) radioGroup.getChildAt(radioGroup.getChildCount() - 1)).setError(null);
        }
    }

    public boolean isRequired() {
        return mRequired;
    }

    public void setRequired(boolean required) {
        this.mRequired = required;
    }

    @Override
    public void serializeToStorage(Map<String, String> storage, String fieldName) {
        int radioButtonID = radioGroup.getCheckedRadioButtonId();
        if (radioButtonID != -1) {
            View radioButton = radioGroup.findViewById(radioButtonID);
            int idx = radioGroup.indexOfChild(radioButton);
            storage.put(fieldName, mConfig.getValues()[idx]);
        } else {
            storage.put(fieldName, "");
        }

    }

    @Override
    public void restoreFromStorage(Map<String, String> storage, String fieldName) {
        String value = storage.get(fieldName);
        for (int i = 0; i < mConfig.getValues().length; i++) {
            if (mConfig.getValues()[i].equals(value)) {
                ((RadioButton) radioGroup.getChildAt(i)).setChecked(true);
                break;
            }
        }
    }
}

File: app/src/main/kotlin/org/bspb/smartbirds/pro/ui/views/SingleChoiceRadioFormInput.kt
package org.bspb.smartbirds.pro.ui.views

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.*
import org.bspb.smartbirds.pro.R
import org.bspb.smartbirds.pro.backend.dto.Nomenclature
import org.bspb.smartbirds.pro.ui.exception.ViewValidationException
import org.bspb.smartbirds.pro.ui.utils.Configuration
import org.bspb.smartbirds.pro.ui.utils.NomenclaturesBean
import org.bspb.smartbirds.pro.ui.utils.NomenclaturesBean_
import org.bspb.smartbirds.pro.ui.views.SingleChoiceFormInput.NomenclatureItem
import org.bspb.smartbirds.pro.utils.inflate

class SingleChoiceRadioFormInput : FrameLayout, SupportRequiredView, SupportStorage {


    private var mValues: List<Nomenclature>? = null
    private var mRequired = false
    private var mEntriesKey: CharSequence? = null
    private var mHint: CharSequence? = null
    private var mIsVertical = false


    private lateinit var hintTextView: TextView
    private lateinit var radioGroup: RadioGroup

    private var mSelectedItem: NomenclatureItem? = null

    private var nomenclatures: NomenclaturesBean = NomenclaturesBean_.getInstance_(context)

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(
        context,
        attrs,
        R.attr.singleChoiceRadioFormInputStyle
    )

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        val a = context.obtainStyledAttributes(
            attrs,
            R.styleable.SingleChoiceRadioFormInput,
            defStyleAttr,
            0
        )
        try {
            mHint = a.getText(R.styleable.SingleChoiceRadioFormInput_hint)
            mEntriesKey =
                a.getText(R.styleable.SingleChoiceRadioFormInput_entriesType)
            mRequired = a.getBoolean(R.styleable.SingleChoiceRadioFormInput_required, false)
            mIsVertical =
                a.getInteger(R.styleable.SingleChoiceRadioFormInput_orientation, 0) == 1
        } finally {
            a.recycle()
        }

        initView()
    }

    private fun initView() {
        inflate(R.layout.single_choice_radio, true)
        hintTextView = findViewById(R.id.single_choice_radio_hint)
        radioGroup = findViewById(R.id.single_choice_radio_group)

        if (mIsVertical) {
            radioGroup.orientation = LinearLayout.VERTICAL
        }

        radioGroup.setOnCheckedChangeListener { group, checkedId ->
            if (checkedId != -1) {
                hintTextView.error = null
            }

            if (checkedId == -1) {
                mSelectedItem = null
            } else {
                val radioButton = radioGroup.findViewById<View>(checkedId)
                val idx = radioGroup.indexOfChild(radioButton)
                mValues?.also {
                    mSelectedItem = NomenclatureItem(it[idx])
                }
            }
        }

        mHint?.apply {
            if (isNotEmpty()) {
                hintTextView.text = mHint.toString() + ":"
            }
        }

        require(mEntriesKey != null && mEntriesKey!!.isNotEmpty()) { "Entries are required for this form input." }

        mValues = nomenclatures.getNomenclature(mEntriesKey.toString())

        mValues?.also {
            for (item in it) {
                val radioButton = RadioButton(context)
                radioButton.text = item.label.get(context.getString(R.string.locale))
                radioGroup.addView(radioButton)
            }
        }
    }

    override fun checkRequired() {
        if (mRequired && isEnabled) {
            if (radioGroup.checkedRadioButtonId == -1) {
                hintTextView.error = context.getString(R.string.required_field)
                throw ViewValidationException()
            } else {
                hintTextView.error = null
            }
        } else {
            hintTextView.error = null
        }
    }

    override fun serializeToStorage(storage: MutableMap<String, String>, fieldName: String) {
        storage[fieldName] =
            mSelectedItem?.label.toString().replace("\n", Configuration.MULTIPLE_CHOICE_DELIMITER)
        val locale = context.getString(R.string.locale)
        if (mSelectedItem != null) {
            if (mSelectedItem!!.nomenclature.label.hasValue(locale)) {
                storage["$fieldName.$locale"] = mSelectedItem!!.nomenclature.label[locale]
            }
            storage["$fieldName.en"] = mSelectedItem!!.nomenclature.label["en"]
        } else {
            storage["$fieldName.$locale"] = ""
            storage["$fieldName.en"] = ""
        }
    }

    override fun restoreFromStorage(storage: MutableMap<String, String>, fieldName: String) {
        val value = storage["$fieldName.${Configuration.NOMENCLATURE_ID_LANGUAGE}"]
        mValues?.forEachIndexed { index, nomenclature ->
            if (nomenclature.label[Configuration.NOMENCLATURE_ID_LANGUAGE] == value) {
                (radioGroup.getChildAt(index) as RadioButton).isChecked = true
                mSelectedItem = NomenclatureItem(nomenclature)
            }
        }
    }
}